jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+            if (subprotocol!=null)
+                baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL,subprotocol);
+            baseResponse.flushBuffer();
+            baseRequest.setHandled(true);
+
+            // upgrade
+            if (LOG.isDebugEnabled())
+                LOG.debug("upgrade connection={} session={}",connection,session);
+
+            baseRequest.setAttribute(HttpConnection.UPGRADE_CONNECTION_ATTRIBUTE, connection);
+            return true;
+        }
+    };
+
+
+
+    static class TestWebSocketConnectionFactory extends ContainerLifeCycle implements WebSocketConnectionFactory
+    {
+        WebSocketPolicy policy = new WebSocketPolicy(WebSocketBehavior.SERVER);
+        ByteBufferPool bufferPool = new MappedByteBufferPool();
+        WebSocketExtensionRegistry extensionRegistry = new WebSocketExtensionRegistry();
+        DecoratedObjectFactory objectFactory = new DecoratedObjectFactory();
+
+        @Override
+        public WebSocketCoreConnection newConnection(Connector connector, EndPoint endPoint, List<String> extensions)
+        {
+            ExtensionStack extensionStack = new ExtensionStack(extensionRegistry);
+            extensionStack.negotiate(objectFactory, policy, bufferPool,
+                                     extensions.stream().map(ExtensionConfig::parse).collect(Collectors.toList()));
+
+            UpgradeRequest upgradeRequest = null; // TODO
+            UpgradeResponse upgradeResponse = null; // TODO
+
+            return new WebSocketCoreConnection(
+                    endPoint,
+                    connector.getExecutor(),
+                    bufferPool,
+                    objectFactory,
+                    policy,
+                    extensionStack,
+                    upgradeRequest,
+                    upgradeResponse);
+        }
+    }
+
+
+    static class TestWebSocketSessionFactory extends ContainerLifeCycle implements WebSocketSessionFactory
+    {
+        @Override
+        public WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint>
+        newSession(WebSocketCoreConnection connection, String subprotocol)
+        {
+
+            // TODO why is remoteEndpoint an interface rather than just an impl?
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
+            remoteEndpoint.open(); // TODO why is this needed?
+
+            // TODO abstract the creation of a local Endpoint
+            WebSocketLocalEndpoint localEndpoint = new WebSocketLocalEndpoint()
+            {
+                @Override
+                public Logger getLog()
+                {
+                    return LOG;
+                }
+
+                @Override
+                public boolean isOpen()
+                {
+                    return connection.isOpen();
+                }
+
+                @Override
+                public void onOpen()
```

Comment:
Where's your Session?

Reply:
In this case, my test "API" layer has no requirement on the session beyond what is provided by the core, so the core session is just fine and does not need to be extended/wrapped/mapped etc.  In fact the core session already does too much.